README.rst
Introduction
============

This recipe adds additional ``filestorage`` and ``zodb_db`` stanzas to the ``zope.conf``
and ``zeo.conf`` files generated by the ``plone.recipe.zope2instance`` and
``plone.recipe.[zope2]zeoserver`` recipes.  It also creates the directories in which
the extra filestorage data files will be created.  This makes it easy to add
additional filestorages and mountpoints to a Zope 5 instance.


Supported options
=================

parts
    A list of filestorage sub-parts to be generated, one per line.  (This is
    different from buildout parts.)
zeo
    The name of a ``plone.recipe.zope2zeoserver`` or ``plone.recipe.zeoserver`` part to
    which we want to add the extra filestorage.  Defaults to the first such part
    in the buildout, if any.
zopes
    A list of names of ``plone.recipe.zope2instance`` parts to which we want to add
    the extra filestorage.  Defaults to all ``plone.recipe.zope2instance`` parts
    connected to the associated ``zeoserver`` part, if any, or all
    ``plone.recipe.zope2instance`` parts, if no ZEO is found.
backup
    The name of a ``collective.recipe.backup`` (>= 2.7) part to which we want
    to add extra filestorage/blobstorage backup entries.

The following options affect the generated ``zope.conf`` and ``zeo.conf``.  Each may be
specified for all filestorage subparts in the ``collective.recipe.filestorage``
buildout part, or for one particular filestorage subpart by placing the option
in a new buildout part called ``filestorage_subpart``, where *subpart* is the name
of the subpart listed in the ``parts`` option of this recipe.  The name of the
subpart may be interpolated by using ``%(fs_part_name)s`` in the option.

allow-implicit-cross-references
    Allow references across mounted databases. This is generally a bad idea.
    Defaults to ``False``.
location
    The location of the ``Data.fs`` file, relative to the buildout root directory.
    Defaults to ``var/filestorage/%(fs_part_name)s/Data.fs``
zodb-name
    The name of the ZODB.  Defaults to ``%(fs_part_name)s``.
zodb-cache-size
    Set the ZODB cache size, i.e. the number of objects which the ZODB cache
    will try to hold.  Inherits from the associated Zope part.  Defaults to
    5000.
zodb-mountpoint
    Set the path to the mountpoint.  Defaults to ``/%(fs_part_name)s``.
    If the object is being mounted at a different path than its location
    in the source database, two paths may be provided using the
    ``virtualpath:realpath`` syntax. For example, when mounting ``/source/bar``
    from ``source.fs`` at ``/foo/bar`` in ``target.fs``:
    ``/foo/bar:/%(fs_part_name)s/bar``. **Note**: the name of the mount point
    and the mounted object must be the same, i.e. ``/foo:/bar`` will fail.
zodb-container-class
    Set the class of the object being mounted.  Defaults to not being set.
zeo-address
    Set the port of the associated ZEO server.  Inherits from the associated
    Zope and ZEO parts.  Defaults to 8100.
zeo-client-cache-size
    Set the size of the ZEO client cache. Inherits from the associated Zope
    part.  Defaults to '30MB'.
zeo-storage
    Set the id of the ZEO storage. Defaults to ``%(fs_part_name)s``.
zeo-client-name
    Set the name of the ZEO client. Defaults to ``%(fs_part_name)s_zeostorage``.
zeo-client-client
    Set the persistent cache name that is used to construct the cache
    filenames. Persistent cache files are disabled by default.
blob-storage
    Set the directory to be used to store blobs for a standalone Zope instance
    or a ZEO server. Optional. Required if you're going to store blobs, though.
    Recommended value: ``var/blobstorage-%(fs_part_name)s``
zeo-blob-storage
    Set the directory to be used to store blobs for a ZEO client. Defaults to
    using the same value as ``blob-storage``.
zeo-shared-blob-dir
    Boolean that should be 'on' if the blob dir is being shared by the ZEO
    server and client. Defaults to 'on'.


Example usage
=============

Here's a minimal buildout that adds an extra filestorage::

   [buildout]
   extends = base.cfg
   parts =
       filestorage
       instance

   [instance]
   recipe = plone.recipe.zope2instance
   user = me

   [filestorage]
   recipe = collective.recipe.filestorage
   parts =
       mystorage

By default, the location of the new filestorage will be:
``var/filestorage/mystorage/mystorage.fs``

See above for options to override the defaults.

A setting can be modified just for one particular filestorage, by creating
a new part with the ``filestorage_`` prefix, like so::

   [filestorage]
   recipe = collective.recipe.filestorage
   parts =
       myfirststorage
       mysecondstorage

   [filestorage_mysecondstorage]
   zodb-cache-size = 1000

By default, the recipe adds the extra filestorages to each
``plone.recipe.zope2instance`` part in the buildout,
but you can tell it to only add it to certain parts::

    [buildout]
    extends = base.cfg
    parts =
        filestorage
        instance1
        instance2

    [instance1]
    recipe = plone.recipe.zope2instance

    [instance2]
    recipe = plone.recipe.zope2instance

    [filestorage]
    recipe = collective.recipe.filestorage
    zopes = instance1
    parts =
        my-fs

Here is a minimal buildout including a ZEO server and two ZODB clients::

    [buildout]
    extends = base.cfg
    parts =
        filestorage
        zeoserver
        primary
        secondary

    [zeoserver]
    recipe = plone.recipe.zope2zeoserver

    [primary]
    recipe = plone.recipe.zope2instance
    zeo-client = on

    [secondary]
    recipe = plone.recipe.zope2instance
    zeo-client = on

    [filestorage]
    recipe = collective.recipe.filestorage
    parts =
        my-fs

As above, we can override a number of the default parameters::

    [buildout]
    extends = base.cfg
    parts =
        filestorage
        zeoserver
        primary
        secondary

    [zeoserver]
    recipe = plone.recipe.zope2zeoserver

    [primary]
    recipe = plone.recipe.zope2instance
    zeo-client = on

    [secondary]
    recipe = plone.recipe.zope2instance
    zeo-client = on

    [filestorage]
    recipe = collective.recipe.filestorage
    location = var/filestorage/%(fs_part_name)s/Data.fs
    blob-storage = var/blobstorage-%(fs_part_name)s
    zodb-cache-size = 1000
    zodb-name = %(fs_part_name)s_db
    zodb-mountpoint = /%(fs_part_name)s_mountpoint
    zeo-address = 8101
    zeo-client-cache-size = 50MB
    zeo-storage = %(fs_part_name)s_storage
    zeo-client-name = %(fs_part_name)s_zeostorage_name
    parts =
        my-fs

By default, the recipe adds the extra filestorages to the first
``plone.recipe.zope2zeoserver`` part in the buildout, and will throw an error if
there is more than one part using this recipe.  However, you can override this
behavior by specifying a particular ZEO part.  In this case, the filestorages
will only be added to the Zopes using that ZEO, by default::

    [buildout]
    extends = base.cfg
    parts =
        filestorage
        zeoserver1
        zeoserver2
        primary
        secondary
        other-zope

    [zeoserver1]
    recipe = plone.recipe.zope2zeoserver
    zeo-address = 8100

    [zeoserver2]
    recipe = plone.recipe.zope2zeoserver
    zeo-address = 8101

    [primary]
    recipe = plone.recipe.zope2instance
    zeo-client = 1
    zeo-address = 8101

    [secondary]
    recipe = plone.recipe.zope2instance
    zeo-client = 1
    zeo-address = 8101

    [other-zope]
    recipe = plone.recipe.zope2instance
    zeo-client = 1
    zeo-address = 8100

    [filestorage]
    recipe = collective.recipe.filestorage
    zeo = zeoserver2
    parts =
        my-fs


Backup integration
==================

Here's a buildout that illustrates backup integration::

    [buildout]
    extends = base.cfg
    parts =
        filestorage
        instance
        backup

    [instance]
    recipe = plone.recipe.zope2instance
    user = me:pass

    [backup]
    recipe = collective.recipe.backup>=2.7
    
    [filestorage]
    recipe = collective.recipe.filestorage
    parts =
        foo
        bar
    backup = backup


Running the tests
=================

The github checkout of ``collective.recipe.filestorage`` includes a buildout
which installs a script for running the tests. For this to work, you need to
have the test dependencies installed::

    python bootstrap.py
    bin/buildout
    bin/test

Alternatively, you can change the final step to ``./bin/python setup.py test``
which will retrieve the test dependencies and run the tests.

Known issue: The tests run buildout in a separate process, so it's currently
impossible to put a ``pdb`` breakpoint in the recipe and debug during the test.
If you need to do this, set up another buildout which uses
``collective.recipe.filestorage``
as a development egg.


Reporting bugs or asking questions
==================================

Use the github tracker:
https://github.com/collective/collective.recipe.filestorage/issues

Some old bugs are at Launchpad:
https://bugs.launchpad.net/collective.buildout/


CHANGES.txt
0.7 (unreleased)
================

 - Fix test for python3 version
   [aormazabal]
 - Add support for c.r.backup
   [mamico]
 - Document virtual mount points. [jean]
 - Disallow cross-database references finally (see https://dev.plone.org/ticket/11150)
   [jean]

0.7 (unlreased)
===============

- Support zrs options replicate-to, replicate-from and keep-alive-delay
  [ivant, djay]


0.6 (2010-08-13)
================

 - Support use of the recipe when the ZEO server and client are in different
   buildouts.
   [ramon]

0.5 (2010-02-03)
================

 - Support plone.recipe.zeoserver in addition to plone.recipe.zope2zeoserver.
   [davisagli]

0.4 (2009-12-30)
================

 - Add support for creating blob storages associated with the created
   filestorage parts.
   [davisagli]

 - Fixed message indicating which parts the filestorage part must be installed
   before.  This fixes https://bugs.launchpad.net/bugs/433877 (Thanks, Jean
   Jordaan)
   [davisagli]

0.3 (2009-03-19)
================

 - Avoid causing the installation of all parts, even those not listed in the
   [buildout] parts variable.  Thanks to Dylan Jay for the diagnosis and fix.
   [davisagli]

 - Fixed restructured text of long description (README.txt).
   [maurits]

0.2 (2008-04-30)
================

 - Added zodb_container_class option.
   [davisagli]

0.1 (2008-04-22)
================

 - Happy Earth Day!
   [davisagli]

 - Created recipe with ZopeSkel.  Initial implementation.
   [davisagli]


collective/recipe/filestorage/tests/test_docs.py
# -*- coding: utf-8 -*-
"""
Doctest runner for 'collective.recipe.filestorage'.
"""
__docformat__ = 'restructuredtext'

import os
import re
import unittest
import doctest
import zc.buildout.tests
import zc.buildout.testing

from zope.testing import renormalizing

optionflags = (doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE | doctest.REPORT_ONLY_FIRST_FAILURE)

current_dir = os.path.abspath(os.path.dirname(__file__))
recipe_location = current_dir
zope2_location = os.path.join(current_dir, 'zope2')


def setUp(test):
    zc.buildout.testing.buildoutSetUp(test)

    # Install any other recipes that should be available in the tests
    zc.buildout.testing.install('plone.recipe.zope2instance', test)
    zc.buildout.testing.install('zc.recipe.egg', test)

    # Install the recipe in develop mode
    zc.buildout.testing.install_develop('collective.recipe.filestorage', test)

    # Add a base.cfg we can extend
    zc.buildout.testing.write('base.cfg', '''
[buildout]
extends = https://dist.plone.org/release/5.2.9/versions.cfg
index = https://pypi.org/simple
[versions]
# pin to a version that doesn't pull in an eggified Zope
waitress = 2.0.0
''')


def test_suite():
    suite = unittest.TestSuite((
        doctest.DocFileSuite(
            'doctests.rst',
            setUp=setUp,
            tearDown=zc.buildout.testing.buildoutTearDown,
            optionflags=optionflags,
            checker=renormalizing.RENormalizing([
                # ignore warnings in output
                (re.compile('^.*?WARNING:.*?$', re.M), ''),
                (re.compile('^.*?warning:.*?$', re.M), ''),
                (re.compile('^.*?zip_safe.*?$', re.M), ''),
                (re.compile('^.*?module references __path__.*?$', re.M), ''),
                zc.buildout.testing.normalize_path,
            ]),
            globs=globals(),
        ),
    ))
    return suite


if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


collective/recipe/filestorage/tests/zope2/lib/python/ZEO/mkzeoinst.py
# -*- coding: utf-8 -*-
import os

join = os.path.join


class ZEOInstanceBuilder:

    def create(self, location, *args):
        os.mkdir(location)
        for sub in ('etc', 'bin'):
            os.mkdir(join(location, sub))
        content = 'PYTHONPATH="$ZODB3_HOME"'
        bin_dir = join(location, 'bin')
        for script in ('runzeo', 'zeoctl'):
            f = open(join(bin_dir, script), 'w')
            f.write(content)
            f.close()


collective/recipe/filestorage/__init__.py
# -*- coding: utf-8 -*-
"""Recipe filestorage"""

import os

from zc.buildout import UserError


class Recipe(object):
    """zc.buildout recipe"""

    def __init__(self, buildout, name, options):
        self.buildout, self.name, self.options = buildout, name, options
        active_parts = [p.strip() for p in self.buildout['buildout']['parts'].split()]

        # figure out which ZEO we're going to inject filestorage configuration into, if any
        zeo_address = None
        self.zeo_part = options.get('zeo', None)
        if self.zeo_part is not None:
            if self.zeo_part in self.buildout:
                zeo_address = self.buildout[self.zeo_part].get('zeo-address', 8100)
            else:
                raise UserError('[collective.recipe.filestorage] "%s" '
                                'part specifies nonexistant zeo part "%s".' % (name, self.zeo_part))
        else:
            for part_name in active_parts:
                part = self.buildout[part_name]
                if 'recipe' not in part:
                    continue
                elif part['recipe'] in ('plone.recipe.zope2zeoserver', 'plone.recipe.zeoserver'):
                    if self.zeo_part is not None:
                        raise UserError('[collective.recipe.filestorage] "%s" part found multiple zeoserver parts; '
                                        'please specify which one to use with the "zeo" option.' % name)
                    self.zeo_part = part_name
                    zeo_address = part.get('zeo-address', 8100)

        # figure out which Zopes we're going to inject filestorage configuration into
        self.zope_parts = options.get('zopes', '').split()
        if len(self.zope_parts) == 0:
            for part_name in active_parts:
                part = self.buildout[part_name]
                if 'recipe' not in part:
                    continue
                elif part['recipe'] == 'plone.recipe.zope2instance':
                    if zeo_address is None or zeo_address == part.get('zeo-address', 8100):
                        self.zope_parts.append(part_name)

        # figure out c.r.backup recipe
        self.backup_part = options.get('backup', None)
        if self.backup_part is not None:
            if self.backup_part not in self.buildout:
                raise UserError('[collective.recipe.filestorage] "%s" part specifies '
                                'nonexistant backup part "%s".' % (name, self.backup_part))

        # make sure this part is before any associated zeo/zope parts in the
        # buildout parts list
        self._validate_part_order()

        # inject the extra sections into the correct zope-conf-additional or
        # zeo-conf-additional variables.
        self.subparts = options.get('parts', '').split()
        for subpart in self.subparts:
            for zope_part in self.zope_parts:
                self._inject_zope_conf(zope_part, subpart)
            if self.zeo_part is not None:
                self._inject_zeo_conf(self.zeo_part, subpart)
            if self.backup_part is not None:
                self._inject_backup_additional(self.backup_part, subpart)

    def install(self):

        for subpart in self.subparts:
            # create the directory for this filestorage
            location = self._subpart_option(subpart, 'location',
                                            default=os.path.join('var', 'filestorage', '%(fs_part_name)s',
                                                                 '%(fs_part_name)s.fs'))
            location = os.path.join(self.buildout['buildout']['directory'], location)
            fs_dir = os.path.dirname(location)
            if not os.path.exists(fs_dir):
                os.makedirs(fs_dir)

            # create blobstorage dirs
            blob_storage = os.path.join('var', 'blobstorage-%(fs_part_name)s')
            if self._subpart_option(subpart, 'blob-storage', default=''):
                blob_storage = self._subpart_option(subpart, 'blob-storage', default=blob_storage)
                if not blob_storage.startswith(os.path.sep):
                    blob_storage = os.path.join(self.buildout['buildout']['directory'], blob_storage)
                if not os.path.exists(blob_storage):
                    os.makedirs(blob_storage)

                if self.zeo_part:
                    zeo_blob_storage = self._subpart_option(subpart, 'zeo-blob-storage', default=blob_storage)
                    if not zeo_blob_storage.startswith(os.path.sep):
                        zeo_blob_storage = os.path.join(self.buildout['buildout']['directory'], zeo_blob_storage)
                    if not os.path.exists(zeo_blob_storage):
                        os.makedirs(zeo_blob_storage)

        # return an empty list because we don't have anything we want buildout to automatically remove
        return tuple()

    def update(self):
        """Updater"""
        pass

    def _validate_part_order(self):
        """ Make sure this part is before any associated zeo/zope parts in the
            buildout parts list.
        """
        injector_parts = [self.name]
        target_parts = self.zope_parts[:]
        if self.zeo_part is not None:
            target_parts.append(self.zeo_part)
        if self.backup_part is not None:
            target_parts.append(self.backup_part)
        for part_name in self.buildout['buildout']['parts'].split():
            if part_name in injector_parts:
                injector_parts.remove(part_name)
                continue
            if part_name in target_parts:
                if len(injector_parts) > 0:
                    raise UserError(
                        '[collective.recipe.filestorage] The "%s" part must be listed '
                        'before the following parts in ${buildout:parts}: %s' % (self.name, ', '.join(target_parts)))
                target_parts.remove(part_name)
        if len(target_parts) > 0:
            raise UserError(
                '[collective.recipe.filestorage] The "%s" part expected but failed '
                'to find the following parts in ${buildout:parts}: %s' % (self.name, ', '.join(target_parts)))

    def _inject_zope_conf(self, zope_part, subpart):
        zope_options = self.buildout[zope_part]

        location = self._subpart_option(subpart, 'location',
                                        default=os.path.join('var', 'filestorage', '%(fs_part_name)s',
                                                             '%(fs_part_name)s.fs'))
        location = os.path.join(self.buildout['buildout']['directory'], location)

        storage_template = file_storage_template
        blob_storage = os.path.join('var', 'blobstorage-%(fs_part_name)s')
        blob_enabled = False
        if self._subpart_option(subpart, 'blob-storage', default=''):
            blob_enabled = True
            blob_storage = self._subpart_option(subpart, 'blob-storage', default=blob_storage)
            if not blob_storage.startswith(os.path.sep):
                blob_storage = os.path.join(self.buildout['buildout']['directory'], blob_storage)
            storage_template = self._blob_storage_template(zope_part)
        storage_snippet = storage_template % dict(
            fs_name='',
            fs_path=location,
            blob_storage=blob_storage,
        )

        if zope_options.get('zeo-client', 'false').lower() in ('yes', 'true', 'on', '1'):
            zeo_address = self._subpart_option(subpart, 'zeo-address', default='8100',
                                               inherit=(zope_part, self.zeo_part))
            zeo_client_cache_size = self._subpart_option(subpart, 'zeo-client-cache-size', default='30MB',
                                                         inherit=zope_part)
            zeo_client_client = self._subpart_option(subpart, 'zeo-client-client', default='', inherit=zope_part)
            if zeo_client_client:
                zeo_client_client = 'client %s' % zeo_client_client
            zeo_storage = self._subpart_option(subpart, 'zeo-storage', default='%(fs_part_name)s')
            zeo_client_name = self._subpart_option(subpart, 'zeo-client-name', default='%(fs_part_name)s_zeostorage')
            zeo_client_var = self._subpart_option(subpart, 'zeo-client-var',
                                                  default=os.path.join(zope_options['location'], 'var'))

            zeo_storage_template = zeo_file_storage_template
            zeo_blob_storage = self._subpart_option(subpart, 'zeo-blob-storage', default=blob_storage)
            if not zeo_blob_storage.startswith(os.path.sep):
                zeo_blob_storage = os.path.join(self.buildout['buildout']['directory'], zeo_blob_storage)
            zeo_shared_blob_dir = self._subpart_option(subpart, 'zeo-shared-blob-dir', default='on')
            if blob_enabled:
                zeo_storage_template = zeo_blob_storage_template

            storage_snippet = zeo_storage_template % dict(
                zeo_address=zeo_address,
                zeo_client_cache_size=zeo_client_cache_size,
                zeo_client_client=zeo_client_client,
                zeo_storage=zeo_storage,
                zeo_client_name=zeo_client_name,
                zeo_client_var=zeo_client_var,
                zeo_blob_storage=zeo_blob_storage,
                zeo_shared_blob_dir=zeo_shared_blob_dir,
            )

        zodb_cache_size = self._subpart_option(subpart, 'zodb-cache-size', default='5000', inherit=zope_part)
        allow_implicit_cross_references = self._subpart_option(subpart, 'allow-implicit-cross-references',
                                                               default='false')
        zodb_name = self._subpart_option(subpart, 'zodb-name', default='%(fs_part_name)s')
        zodb_mountpoint = self._subpart_option(subpart, 'zodb-mountpoint', default='/%(fs_part_name)s')
        zodb_container_class = self._subpart_option(subpart, 'zodb-container-class', default='')
        if zodb_container_class:
            zodb_container_class = "\n    container-class %s" % zodb_container_class
        zodb_stanza = zodb_template % dict(
            zodb_name=zodb_name,
            allow_implicit_cross_references=allow_implicit_cross_references,
            zodb_mountpoint=zodb_mountpoint,
            zodb_container_class=zodb_container_class,
            zodb_cache_size=zodb_cache_size,
            storage_snippet=storage_snippet.strip(),
        )

        zope_conf_additional = zope_options.get('zope-conf-additional', '')
        zope_options['zope-conf-additional'] = zope_conf_additional + zodb_stanza

    def _inject_zeo_conf(self, zeo_part, subpart):
        zeo_options = self.buildout[zeo_part]

        location = self._subpart_option(subpart, 'location',
                                        default=os.path.join('var', 'filestorage', '%(fs_part_name)s',
                                                             '%(fs_part_name)s.fs'))
        location = os.path.join(self.buildout['buildout']['directory'], location)
        zeo_storage = self._subpart_option(subpart, 'zeo-storage', default='%(fs_part_name)s')

        storage_template = file_storage_template
        blob_storage = os.path.join('var', 'blobstorage-%(fs_part_name)s')
        if self._subpart_option(subpart, 'blob-storage', default=''):
            blob_storage = self._subpart_option(subpart, 'blob-storage', default=blob_storage)
            if not blob_storage.startswith(os.path.sep):
                blob_storage = os.path.join(self.buildout['buildout']['directory'], blob_storage)
            storage_template = self._blob_storage_template(zeo_part)

        storage_snippet = storage_template % dict(
            fs_name=zeo_storage,
            fs_path=location,
            blob_storage=blob_storage,
        )

        zeo_conf_additional = zeo_options.get('zeo-conf-additional', '')
        zeo_options['zeo-conf-additional'] = zeo_conf_additional + storage_snippet

    def _inject_backup_additional(self, backup_part, subpart):
        backup_options = self.buildout[backup_part]

        location = self._subpart_option(subpart, 'location',
                                        default=os.path.join('var', 'filestorage', '%(fs_part_name)s',
                                                             '%(fs_part_name)s.fs'))
        location = os.path.join(self.buildout['buildout']['directory'], location)
        zeo_storage = self._subpart_option(subpart, 'zeo-storage', default='%(fs_part_name)s')
        backup_template = "%(fs_name)s %(fs_path)s"
        blob_storage = os.path.join('var', 'blobstorage-%(fs_part_name)s')
        if self._subpart_option(subpart, 'blob-storage', default=''):
            blob_storage = self._subpart_option(subpart, 'blob-storage', default=blob_storage)
            if not blob_storage.startswith(os.path.sep):
                blob_storage = os.path.join(self.buildout['buildout']['directory'], blob_storage)
            backup_template = "%(fs_name)s %(fs_path)s %(blob_storage)s"

        additional = backup_template % dict(
            fs_name=zeo_storage,
            fs_path=location,
            blob_storage=blob_storage,
        )

        backup_additionals = backup_options.get('additional_filestorages', '')
        backup_options['additional_filestorages'] = '\n'.join([backup_additionals, additional])

    def _subpart_option(self, subpart, option, default=None, inherit=()):
        """
        Retrieve an option for a filestorage subpart, perhaps falling back to other specified parts.
        Also substitutes the name of the subpart.
        """
        parts_to_check = ['filestorage_' + subpart, self.name]
        if type(inherit) == type(''): # noqa
            inherit = (inherit,)
        parts_to_check.extend(inherit)

        val = default
        for part in parts_to_check:
            if part not in self.buildout:
                continue
            if option in self.buildout[part]:
                val = self.buildout[part][option]
                break

        return val % dict(
            fs_part_name=subpart,
        )

    def _blob_storage_template(self, part):
        if 'zope2-location' in self.buildout[part]:
            # non-eggified Zope; assume ZODB 3.8.x
            return blob_storage_zodb_3_8_template
        else:
            return blob_storage_zodb_3_9_template


# Storage snippets for zope.conf template
file_storage_template = """
    <filestorage %(fs_name)s>
      path %(fs_path)s
    </filestorage>
"""

blob_storage_zodb_3_8_template = """
    <blobstorage %(fs_name)s>
      blob-dir %(blob_storage)s
      <filestorage %(fs_name)s>
        path %(fs_path)s
      </filestorage>
    </blobstorage>
"""

blob_storage_zodb_3_9_template = """
    <filestorage %(fs_name)s>
      path %(fs_path)s
      blob-dir %(blob_storage)s
    </filestorage>
"""

zeo_file_storage_template = """
    <zeoclient>
      server %(zeo_address)s
      storage %(zeo_storage)s
      name %(zeo_client_name)s
      var %(zeo_client_var)s
      cache-size %(zeo_client_cache_size)s
      %(zeo_client_client)s
    </zeoclient>
""".strip()

zeo_blob_storage_template = """
    <zeoclient>
      blob-dir %(zeo_blob_storage)s
      shared-blob-dir %(zeo_shared_blob_dir)s
      server %(zeo_address)s
      storage %(zeo_storage)s
      name %(zeo_client_name)s
      var %(zeo_client_var)s
      cache-size %(zeo_client_cache_size)s
      %(zeo_client_client)s
    </zeoclient>
""".strip()

zodb_template = """
<zodb_db %(zodb_name)s>
    cache-size %(zodb_cache_size)s
    allow-implicit-cross-references %(allow_implicit_cross_references)s
    %(storage_snippet)s
    mount-point %(zodb_mountpoint)s%(zodb_container_class)s
</zodb_db>
"""
